enabling the CDN

qx-will opened the pull request:
Enabling the CDN resource for web app.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/yaronyg/sleeveforarm/31)
<!-- Reviewable:end -->

qixwang committed:
enabling the CDN

yaronyg commented:
@qx-will Please see inline comments. Thanks!

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/serviceEnvironment.ts, line 36 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KutpwcQ2xb1L3yAj3T0:-KutpwcR4o7tRGGBh42f:b-wqg2o2) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/serviceEnvironment.ts#L36)):*
> ```XML
> }
> 
> export function getCDNEndpoint() {
> ```

Let's imagine a user is writing their web site and they program it to use the CDN to return some URLs and not use the CDN for other URLs. Now they are running the code in their development environment. We don't deploy the CDN there. So what happens to their code? It's going to call this API and get back null. Isn't that going to break code paths and tests while the developer is developing their site using Sleeve?

I think we need to return something reasonable in all situations, not just production. I'm thinking we could return http://127.0.0.1/. It's not great but it's something. So if ServiceEnvironmentUtilities.cdnprefix doesn't return anything then we default to http://127.0.0.1.

This brings up an obvious problem though. How do we handle HTTP vs HTTPS? Do we need two environment variables? One for each endpoint? Or should we just return the domain and let the dev append http or https depending on their preference?

---

*[src/webapp-node-azureInfrastructure.ts, line 189 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu1IuzwkVDG7gBbJK5:-Kuu1IuzwkVDG7gBbJK6:bxcni79) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/webapp-node-azureInfrastructure.ts#L189)):*
> ```XML
>         await this.deployToWebApp(developmentDeploy);
> 
>         if (this.DefaultCDNSKU !== undefined) {
> ```

The deployToProduction method is getting crazy long. We need to factor the CDN code out into its own method and just call it from deployToProduction. Also, in order not to slow everything down we should make sure that the method is async and we should push it onto secondStepPromises. That way the CDN setup can run in parallel with other commands like 'az webapp config appsettings set' above as well as 'az webapp deployment source config-local-git'.

---

*[test/webapp-node-azureTest.ts, line 51 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu28MVkGet2D5bi-PF:-Kuu28MVkGet2D5bi-PG:bhh88jy) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/test/webapp-node-azureTest.ts#L51)):*
> ```XML
>             to: "const option = require(\"sleeveforarm/src/webapp-node-azure\").CDNSKUOption; module.exports = new webappNodeAzure().setCDNProvider(option.Standard_Akamai);"
>         };
>         // tslint:disable-next-line:max-line-length
> ```

Why is this tslint command here? The next line is empty.

---

*[test/webapp-node-azureTest.ts, line 54 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu2BTr3JVGLzXufG7j:-Kuu2BTsYFIMxVq_0jJz:b-pvve4h) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/test/webapp-node-azureTest.ts#L54)):*
> ```XML
> 
>         await ReplaceInFile(replaceInFileOptions);
>         // await ReplaceInFile(replaceInFileOptions);
> ```

Why is this quoted line here? Isn't it just replacing the line above it?

---

*[testAssets/index.ts, line 48 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu2N7y0vDRphZcreiz:-Kuu2N7y0vDRphZcrej-:b-144lya) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/testAssets/index.ts#L48)):*
> ```XML
>     const needReDirect = request.url.indexOf("cdn") > 0;
> 
>     if (needReDirect) {
> ```

It would be good to put a quick comment here explaining how we are using this to test the CDN.

---

*[testAssets/index.ts, line 50 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu2JCb9Qt3WBthkaWn:-Kuu2JCb9Qt3WBthkaWo:bqggmwx) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/testAssets/index.ts#L50)):*
> ```XML
>     if (needReDirect) {
>          const urlToRedirect = ServiceEnvironment.getCDNEndpoint();
>         // tslint:disable-next-line:max-line-length
> ```

Why is this tlint line here? The next line isn't that long.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-:-Kuu2rt534Fj8Q-8WEbR:bnamx6z)*
<!-- Sent from Reviewable.io -->

qx-will commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/serviceEnvironment.ts, line 36 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KutpwcQ2xb1L3yAj3T0:-Kv5BRtayZ3y_pYiMIV9:b9cvvw) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/serviceEnvironment.ts#L36)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

Let's imagine a user is writing their web site and they program it to use the CDN to return some URLs and not use the CDN for other URLs. Now they are running the code in their development environment. We don't deploy the CDN there. So what happens to their code? It's going to call this API and get back null. Isn't that going to break code paths and tests while the developer is developing their site using Sleeve?

I think we need to return something reasonable in all situations, not just production. I'm thinking we could return http://127.0.0.1/. It's not great but it's something. So if ServiceEnvironmentUtilities.cdnprefix doesn't return anything then we default to http://127.0.0.1.

This brings up an obvious problem though. How do we handle HTTP vs HTTPS? Do we need two environment variables? One for each endpoint? Or should we just return the domain and let the dev append http or https depending on their preference?
</blockquote></details>

how about solving this in a simpler way. i put a check before the test of CDN, if the CDN feature is not enabled, then skip this test, this should fit into the dev machine.

---

*[src/webapp-node-azureInfrastructure.ts, line 189 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu1IuzwkVDG7gBbJK5:-Kv5DGskts6nTlJpxV91:b-rmvykb) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/webapp-node-azureInfrastructure.ts#L189)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

The deployToProduction method is getting crazy long. We need to factor the CDN code out into its own method and just call it from deployToProduction. Also, in order not to slow everything down we should make sure that the method is async and we should push it onto secondStepPromises. That way the CDN setup can run in parallel with other commands like 'az webapp config appsettings set' above as well as 'az webapp deployment source config-local-git'.
</blockquote></details>

done, have a separate method to kick off the az commond for CDN. but thoese three methods should be executed in sequence, so could we still use promise? 

---

*[test/webapp-node-azureTest.ts, line 51 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu28MVkGet2D5bi-PF:-Kv5Dd4CyvsK-vgErAFI:b-7ql1td) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/test/webapp-node-azureTest.ts#L51)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

Why is this tslint command here? The next line is empty.
</blockquote></details>

fixed

---

*[test/webapp-node-azureTest.ts, line 54 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu2BTr3JVGLzXufG7j:-Kv5DeaLRDNYtMRIw-Qb:b-7ql1td) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/test/webapp-node-azureTest.ts#L54)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

Why is this quoted line here? Isn't it just replacing the line above it?
</blockquote></details>

fixed

---

*[testAssets/index.ts, line 48 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu2N7y0vDRphZcreiz:-Kv5Dg4R6edKTqFCDUOH:b-7rl0jd) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/testAssets/index.ts#L48)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

It would be good to put a quick comment here explaining how we are using this to test the CDN.
</blockquote></details>

done.

---

*[testAssets/index.ts, line 50 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu2JCb9Qt3WBthkaWn:-Kv5DhPmgJkuWHI628si:b-7ql1td) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/testAssets/index.ts#L50)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

Why is this tlint line here? The next line isn't that long.
</blockquote></details>

fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

qixwang committed:
update code after 1rd view

yaronyg committed:
Merge remote-tracking branch 'origin' into enableCDNIssue28Merge

# Conflicts:
#	src/serviceEnvironmentUtilities.ts

yaronyg commented:
Review status: 3 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/serviceEnvironment.ts, line 36 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KutpwcQ2xb1L3yAj3T0:-Kv8RKdqljKlLL9HlYWz:bp76qq0) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/serviceEnvironment.ts#L36)):*
<details><summary><i>Previously, qx-will (QX-Will) wrote…</i></summary><blockquote>

how about solving this in a simpler way. i put a check before the test of CDN, if the CDN feature is not enabled, then skip this test, this should fit into the dev machine.
</blockquote></details>

1. If the person calls getCDNEndpoint and no CDN is configured then we need to return an error and tell the user how to configure a CDN (we can file a bug on that later one)

2.In the case that we are in a dev deployment we need to return localhost

---

*[src/webapp-node-azureInfrastructure.ts, line 189 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu1IuzwkVDG7gBbJK5:-Kv8TBmh363CS2hb54rw:b-o2so8t) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/webapp-node-azureInfrastructure.ts#L189)):*
<details><summary><i>Previously, qx-will (QX-Will) wrote…</i></summary><blockquote>

done, have a separate method to kick off the az commond for CDN. but thoese three methods should be executed in sequence, so could we still use promise? 
</blockquote></details>

Please see our comments about how to make the CDN run in parallel from our chat

---

*[src/webapp-node-azureInfrastructure.ts, line 102 at r2](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kv8STm97J3r_04Vw9Fn:-Kv8STm97J3r_04Vw9Fo:bnp2uly) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/d5bf4a1ea1cd5fe23ddc47ebae4de0efa6e8a098/src/webapp-node-azureInfrastructure.ts#L102)):*
> ```XML
>     }
> 
>     public isCDNEnabled(): boolean {
> ```

I think we can remove this function

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

qixwang committed:
Merge pull request #35 from yaronyg/enableCDNIssue28Merge

Enable cdn issue28 merge

qixwang committed:
submit the change from merge of AppInsight

qx-will commented:
Review status: 1 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/serviceEnvironment.ts, line 36 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KutpwcQ2xb1L3yAj3T0:-KvBpEtXoBMMkJWHK9rI:b41k44z) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/serviceEnvironment.ts#L36)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

1. If the person calls getCDNEndpoint and no CDN is configured then we need to return an error and tell the user how to configure a CDN (we can file a bug on that later one)

2.In the case that we are in a dev deployment we need to return localhost
</blockquote></details>

fixed by returning 127.0.0.1 if CDN is not available

---

*[src/webapp-node-azureInfrastructure.ts, line 189 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu1IuzwkVDG7gBbJK5:-KvBpTQKw-RHFD7qaTPN:btptckj) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/webapp-node-azureInfrastructure.ts#L189)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

Please see our comments about how to make the CDN run in parallel from our chat
</blockquote></details>

fixed . 

---

*[src/webapp-node-azureInfrastructure.ts, line 102 at r2](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kv8STm97J3r_04Vw9Fn:-KvBpO4AhrmnXbwrBJCm:b-7ql1td) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/d5bf4a1ea1cd5fe23ddc47ebae4de0efa6e8a098/src/webapp-node-azureInfrastructure.ts#L102)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

I think we can remove this function
</blockquote></details>

fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

yaronyg commented:
Reviewed 1 of 3 files at r2, 1 of 5 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[.gitignore, line 80 at r4](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KvDXg0u-gY5chNlEzr7:-KvDXg0u-gY5chNlEzr8:b-4q8gnf) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/e561c6387a32a33e374d98c54857deb4150eb035/.gitignore#L80)):*
> ```
> 
> # Where we create testing related temporary data
> disposableTestFiles
> ```

You don't need to add the sub-directories. You can just specify ".vs" and that will ignore the directory and all of its contents.

---

*[src/serviceEnvironment.ts, line 36 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KutpwcQ2xb1L3yAj3T0:-KvDWGrADR9ukIySXtYt:b-salz61) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/serviceEnvironment.ts#L36)):*
<details><summary><i>Previously, qx-will (QX-Will) wrote…</i></summary><blockquote>

fixed by returning 127.0.0.1 if CDN is not available
</blockquote></details>

Unfortunately this handles point 2 above but not point 1. We need to return an error if someone calls this function and no CDN actually exists. I think what we need to do is ALWAYS define ServiceenvironmentUtilities.cdnprefx and used a reserved value representing "NO CDN Defined" that will tell us to return an error if this method is called.

---

*[src/webapp-node-azureInfrastructure.ts, line 189 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu1IuzwkVDG7gBbJK5:-KvDWubmKpGQE5a6vInl:b-a27dfz) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/webapp-node-azureInfrastructure.ts#L189)):*
<details><summary><i>Previously, qx-will (QX-Will) wrote…</i></summary><blockquote>

fixed . 
</blockquote></details>

What you wrote will work but it doesn't actually accomplish the objective of letting us run the CDN set up code mostly in parallel with the webapp code. Think about it, when we defined webAppCreateResult we already awaited the webPromise so by the time we get here all the web app set up (which could have happened in parallel with the CDN set up) has already finished. So there is no point in creating promiseArray at this point, it's too late to get any parallelism.

What we need to do is to create the promise array right after we create webPromise and to stick in the CDN code into the array (if needed) right there. Then await the array (using promise.all) and then we can await AGAIN the webPromise to get webAppCreateResult.

---

*[src/webapp-node-azureInfrastructure.ts, line 205 at r4](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KvDXC6fT-mmHsnzstKU:-KvDXC6fT-mmHsnzstKV:b-2yz4zq) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/e561c6387a32a33e374d98c54857deb4150eb035/src/webapp-node-azureInfrastructure.ts#L205)):*
> ```XML
>         await this.deployToWebApp(developmentDeploy);
> 
>         const promiseArrary = [];
> ```

/promiseArrary/promiseArray/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

qixwang committed:
update after 3rd review

qx-will commented:
Review status: 7 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[.gitignore, line 80 at r4](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KvDXg0u-gY5chNlEzr7:-KvFB8EyPoSZOlHc-bux:bb7lyxf) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/e561c6387a32a33e374d98c54857deb4150eb035/.gitignore#L80)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

You don't need to add the sub-directories. You can just specify ".vs" and that will ignore the directory and all of its contents.
</blockquote></details>

ok. 

---

*[src/serviceEnvironment.ts, line 36 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KutpwcQ2xb1L3yAj3T0:-KvFWBadkbf0D1Q6eYuT:b-q6ag7i) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/serviceEnvironment.ts#L36)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

Unfortunately this handles point 2 above but not point 1. We need to return an error if someone calls this function and no CDN actually exists. I think what we need to do is ALWAYS define ServiceenvironmentUtilities.cdnprefx and used a reserved value representing "NO CDN Defined" that will tell us to return an error if this method is called.
</blockquote></details>

here is what i am try to fix this 
update the getCDNEnopint method to   first to check whether CDNOption is set, if no, then throw  out the error, if already set, then run the below code

 const endpoint = process.env[ServiceEnvironmentUtilities.cdnprefix];
    if (endpoint !== undefined) {
        return endpoint;
    }else {
        return "127.0.0.1";
    }

but there is one issue here, i need to access the DefaultCDNSKU property in the WebappNodeAzure class which is in the file webapp-node-azure.ts file. i could make a public get method in that class, but in the getCDNEnpoint method, how can i make sure i am accessing the same runtime object that the user set in the sleeve.js file if he like to enable the CDN? i got stuck at this. 

---

*[src/webapp-node-azureInfrastructure.ts, line 189 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kuu1IuzwkVDG7gBbJK5:-KvBq-XmeK9uipetOPKp:blkwbr5) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/webapp-node-azureInfrastructure.ts#L189)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

What you wrote will work but it doesn't actually accomplish the objective of letting us run the CDN set up code mostly in parallel with the webapp code. Think about it, when we defined webAppCreateResult we already awaited the webPromise so by the time we get here all the web app set up (which could have happened in parallel with the CDN set up) has already finished. So there is no point in creating promiseArray at this point, it's too late to get any parallelism.

What we need to do is to create the promise array right after we create webPromise and to stick in the CDN code into the array (if needed) right there. Then await the array (using promise.all) and then we can await AGAIN the webPromise to get webAppCreateResult.
</blockquote></details>

fixed. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

yaronyg commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[.gitignore, line 80 at r4](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KvDXg0u-gY5chNlEzr7:-KvUkpp06LZlW6_t5lPn:bht8pir) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/e561c6387a32a33e374d98c54857deb4150eb035/.gitignore#L80)):*
<details><summary><i>Previously, qx-will (QX-Will) wrote…</i></summary><blockquote>

ok. 
</blockquote></details>

NB (standard for Non-Blocking and means you can fix it without any follow up from me) - You didn't actually remove the .vs entries and replace with them with a simple .vs

---

*[src/serviceEnvironment.ts, line 36 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KutpwcQ2xb1L3yAj3T0:-KvUl3BjelW0V-dzcq4K:bfzseub) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/serviceEnvironment.ts#L36)):*
<details><summary><i>Previously, qx-will (QX-Will) wrote…</i></summary><blockquote>

here is what i am try to fix this 
update the getCDNEnopint method to   first to check whether CDNOption is set, if no, then throw  out the error, if already set, then run the below code

 const endpoint = process.env[ServiceEnvironmentUtilities.cdnprefix];
    if (endpoint !== undefined) {
        return endpoint;
    }else {
        return "127.0.0.1";
    }

but there is one issue here, i need to access the DefaultCDNSKU property in the WebappNodeAzure class which is in the file webapp-node-azure.ts file. i could make a public get method in that class, but in the getCDNEnpoint method, how can i make sure i am accessing the same runtime object that the user set in the sleeve.js file if he like to enable the CDN? i got stuck at this. 
</blockquote></details>

The problem is that at run time we won't have access to any of the infrastructure files. That is why I suggested that when we are processing the infrastructure files (e.g. after a sleeveforarm deploy command) we always set a value for SErviceEnvironmentUtilities.cdnprefix. If the cdn option isn't set then the value we set is some reserved value like "NO CDN SET". Otherwise we set it to the specified CDN value. That way at run time we will check the environment variable and if we see "NO CDN SET" then we know to return an error to getCDNEndpoint otherwise we just return the specified value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

qixwang committed:
dismiss the CDN option if no setup is available

qx-will commented:
Review status: 5 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/serviceEnvironment.ts, line 36 at r1](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KutpwcQ2xb1L3yAj3T0:-KvacyyESeqrhLZMj-5H:b8o2z0f) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/45d0ae8b23eb0f666731338c2c2c5f53687c3b0b/src/serviceEnvironment.ts#L36)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

The problem is that at run time we won't have access to any of the infrastructure files. That is why I suggested that when we are processing the infrastructure files (e.g. after a sleeveforarm deploy command) we always set a value for SErviceEnvironmentUtilities.cdnprefix. If the cdn option isn't set then the value we set is some reserved value like "NO CDN SET". Otherwise we set it to the specified CDN value. That way at run time we will check the environment variable and if we see "NO CDN SET" then we know to return an error to getCDNEndpoint otherwise we just return the specified value.
</blockquote></details>

push again..:0 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

yaronyg commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/serviceEnvironment.ts, line 37 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KvhVwOeGXXKn-gVlyUB:-KvhVwOfkamgNSfqww4t:bzcoaxq) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/serviceEnvironment.ts#L37)):*
> ```XML
> 
> export function getCDNEndpoint() {
>     if (ServiceEnvironmentUtilities.cdnprefix === "NO_CDN_SET") {
> ```

This check won't work because ServiceEnvironmentUtilities.cdnprefix was set while we were running SleeveForArm. But when this code is run it runs inside of the Node app and is in a completely different environment. So cdnprefix will NEVER equal NO_CDN_SET. See my comments in webapp-node-azureInfrastructure for how to fix this.

---

*[src/serviceEnvironment.ts, line 39 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kvc3ztmXoyvGLuCTAvP:-Kvc3ztmXoyvGLuCTAvQ:b-jyphu6) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/serviceEnvironment.ts#L39)):*
> ```XML
>     if (ServiceEnvironmentUtilities.cdnprefix === "NO_CDN_SET") {
>         throw new Error(
>             `CDN endpoint is not avaialble without proper setup`);
> ```

NB: /avaialble/available/

---

*[src/serviceEnvironment.ts, line 40 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kvc46pxXc_60T4nYDZR:-Kvc46pxXc_60T4nYDZS:b3e3mft) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/serviceEnvironment.ts#L40)):*
> ```XML
>         throw new Error(
>             `CDN endpoint is not avaialble without proper setup`);
>             // why do you even call this method without set up the CDN?
> ```

NB: I think we can remove the commentary :)

---

*[src/serviceEnvironment.ts, line 47 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kvc4CgcusGXdSSqAHZ_:-Kvc4CgdlJi_avn0uM8_:bqx5vj6) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/serviceEnvironment.ts#L47)):*
> ```XML
>         }else {
>             // tslint:disable-next-line:max-line-length
>             return "127.0.0.1"; // very likely this path is under the local dev given the CDN is setup yet no endpoint is avaialbe from the current env property
> ```

It seems like cdnprefix should NEVER be unset. It is either set to NO_CDN_SET or it is set to the proper value. In the case of a dev deployment that proper value is 127.0.0.1. So when we configure a dev deployment we should properly set the CDN value to 127.0.0.1. Which means that if endpoint === undefined then we have a config error and show throw an exception.

See my comment below in webapp-node-azure for more details.

---

*[src/serviceEnvironmentUtilities.ts, line 41 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KvhV-BQQHCTzMfnFBr8:-KvhV-BRhycXNDP4lV5j:b-g6b6of) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/serviceEnvironmentUtilities.ts#L41)):*
> ```XML
> }
> 
> export function dismissCDNOption() {
> ```

I'm pretty sure this function shouldn't be here. serviceEnvironmentUtilities are meant to be run on the user's project, not as part of the SleeveForArm tool.

---

*[src/serviceEnvironmentUtilities.ts, line 49 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kvc5AGqbms-7T7QecW3:-Kvc5AGqbms-7T7QecW4:b9c14z7) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/serviceEnvironmentUtilities.ts#L49)):*
> ```XML
> export const resourcePasswordSuffix = "_password";
> 
> // tslint:disable-next-line:align
> ```

Why would there ben an alignment issue? There is even an auto fix for that in VSCode.

---

*[src/serviceEnvironmentUtilities.ts, line 50 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kvc5FypPIuRtCuw0Wuv:-Kvc5FyqEUpgXqa58MZh:b-y4w391) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/serviceEnvironmentUtilities.ts#L50)):*
> ```XML
> 
> // tslint:disable-next-line:align
> export let cdnprefix = "cdnname";
> ```

Per my suggestions below I think this should be a const.

---

*[src/webapp-node-azureInfrastructure.ts, line 164 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KvhV8gfVTPG822w82Ct:-Kvc70IaF5N1Ft6E3ssW:bvse6ch) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/webapp-node-azureInfrastructure.ts#L164)):*
> ```XML
>             promiseArrary.push(this.enableCDN(webPromise));
>         }else {
>             ServiceEnvironmentUtilities.dismissCDNOption();
> ```

I think we should get rid of dismissCDNOption. Instead I think:
1. serviceEnvironmentUtilities.cdnprefix should be a CONST and should always be equal to "cdnname"
2. I think we should ALWAYS call enableCDN, even if DefaultCDNSKU is undefined. What will happen is that enableCDN (which probably needs a better name) will check DefaultCDNSKU itself and if it is undefined we should set cdnsettingvalue inside of enableCDN to be "NO_CDN_SET". Just like for a dev deployment (as I suggest below) we will skip deploying the CDN but we will still call 'az webapp config appsettings set' to set cdnname="NO_CDN_SET". This just requires changing cdnsettingvalue to "NO_CDN_SET" and having logic that skips the two 'az cdn' commands if the deployment type is DEV or if CDN isn't configured.
3. Then in serviceEnvironment.ts if someone calls getCDNEddpoint we look up ServiceEnvironmentUtilities.cdnprefix (which is always cdnname) and if the value is NO_CDN_SET 

---

*[src/webapp-node-azureInfrastructure.ts, line 215 at r6](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-KvhUgUi1d8KamqwuMBd:-KvhUgUj1V6yodLLUywh:b4q2lm4) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/7aefed74c5b5599a32b98b4da151763b3b996b95/src/webapp-node-azureInfrastructure.ts#L215)):*
> ```XML
> 
>     }
>     private async enableCDN(webPromise: any) {
> ```

What I'm proposing in terms of setting the environment variable for dev deployments is that if CDN is set (e.g. we are calling enableCDN) but if the deployment type is a dev deployment THEN we should obviously skip the 'az cdn profile create' and the 'az cdn endpoint create' commands. But we SHOULD still call 'az webapp config appsettings set' and we should set cdnsettingname using a cdnsettingvalue of "127.0.0.1".

I'm also proposing that we check this.DefaultCDNSKU here and if it is undefined then we will set cdnsettingvalue to "NO_CDN_SET", skip the two az commands as suggested in the previous paragraph, but still call 'az webapp config appsettings' so we set the environment variable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

qixwang committed:
clean up the enableCDN method and remove the dismissCDN method

yaronyg commented:
I like the code! Just one more request, can you please copy the test code from "It should be deployable" into "It should run locally"? If we did everything right then the same code should work the same way with the exception that the CDN requests in local go through 127.0.0.1.

I also opened https://github.com/yaronyg/SleeveForARM/issues/37

---

Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-:-Kw1GPruQqDARXvb4sV5:btth4t5)*
<!-- Sent from Reviewable.io -->

qixwang committed:
update the test code of local test for the CDN part

yaronyg commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[test/webapp-node-azureTest.ts, line 171 at r8](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kw6rC7FlssfNLSgnNCO:-Kw6rC7GVC6nA_N0ZIZ9:bac7pnm) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/2dbcf611b88fdfb6556e3ab3e330fee2a1760e3f/test/webapp-node-azureTest.ts#L171)):*
> ```XML
> 
>           // tslint:disable-next-line:max-line-length
>           // although the CDN is set for the local-test, but we are actually using the 127.0.0.1
> ```

Isn't this the same code we have in the previous test? Can't we factor it out into a function so we don't have to repeat the exact same code?

---

*[test/webapp-node-azureTest.ts, line 202 at r8](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-Kw6qu82EilhqSbiYYeR:-Kw6qu83ioSi52NDYj_y:b-r4mmqt) ([raw file](https://github.com/yaronyg/sleeveforarm/blob/2dbcf611b88fdfb6556e3ab3e330fee2a1760e3f/test/webapp-node-azureTest.ts#L202)):*
> ```XML
> 
>         CommonUtilities.exec("node index.js", fooPath);
>         await getTheResult("http://localhost:1337");
> ```

Don't we need 'await getTheResult(tryCDN);' and related code to prove that things are setup and running right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

qixwang committed:
update the test code to avoid code-duplication

yaronyg commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-:-Kwa6tY7jw8z2FKEOtEi:bnfp4nl)*
<!-- Sent from Reviewable.io -->

qx-will commented:
Reviewed 1 of 3 files at r2, 1 of 5 files at r3, 2 of 6 files at r4, 1 of 4 files at r6, 3 of 3 files at r7, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31)*
<!-- Sent from Reviewable.io -->

yaronyg commented:
I tried to run the tests myself and the local tests failed because the environment variables are not set up right. I have to give a presentation tomorrow about Sleeve so I don't have time to investigate this right now. But I'll look at it after the presentation tomorrow.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/yaronyg/sleeveforarm/31#-:-Kwgl3xU13siUoXv4cge:b-pk5iqg)*
<!-- Sent from Reviewable.io -->

yaronyg committed:
Small fixes

.gitignore - Got rid of individual .vs files and just put out the whole
directory

serviceEnvironment - Whitespace and spelling error

serviceEnvironmentUtilities - The tsline exception wasn't needed
